FLT/ExistsFrobenius.lean

```diff
@@ -0,0 +1,969 @@
+          the ring `A`; this is because the `MulAction` of `L ≃ₐ[K] L` on `Ideal B`
+          is defined using `galRestrict`, which requires reference to the scalars `A`
+          over which the algebraic equivalence `B ≃ₐ[A] B` is defined.
+- `f` : the function assigning a representative non-zero prime ideal of `B` to
+          a coset in the quotient `((L ≃ₐ[K] L) ⧸ decomposition_subgroup_Ideal Q)`,
+          i.e., an element in the orbit of `Q` under the action of `(L ≃ₐ[K] L)`.
+- `p` : a natural prime which is the characteristic of the residue field `(A ⧸ P)`.
+- `q` : the cardinality of the residue field `(A ⧸ P)`. We show that `q = p ^ n` for
+        some `n : ℕ`.
+- `ρ` : `α` in Milne's notation. This `ρ` is an element of `B` which generates the group
+          `(B ⧸ Q)ˣ` and is `0` mod `τ • Q` for any `τ` not in the decomposition subgroup
+          of `Q` over `K`.
+- `α` : local notation for an element `generator : B` which has the properties of `ρ`.
+- `F` : local notation for `F' : B[X]`, a polynomial whose roots are exactly the
+        Galois conjugates of `α`.
+- `m` : local notation for `m' : A[x]`, a polynomial constructed such that its lift to
+        `B` under `algebraMap A B` is equal to `F`.
+- `Frob` : local notation for `Frob' : L ≃ₐ[K] L`, an element of the Galois group that
+           sends `α` to its `q`-th power `mod Q`.
+- `γ` : arbitrary element of `B`. We consider the cases `γ ∈ Q`, `γ ∉ Q`, separately.
+- `i`:  local notation for `i' : ℕ`, such that `γ ≡ α ^ i mod Q`. `i` depends on `γ`.
+
+## Note : I have changed the previous notation `β → α`, `ν → β`, to conform to Milne's notation.
+
+## Implementation notes
+
+This file was written in Prof. Buzzard's 'FLT' repository.
+
+## References
+
+See [Milne (2020)] foonote '2' on p. 141 (p. 143 in PDF)
+  for the proof on which this file is based.
+
+See [Yang & Lutz (2023)], for the definitions of 'AKLB setup' and `galRestrict`.
+
+See [Nakagawa, Baanen, & Filippo (2020)], for the definition of
+  `IsDedekindDomain.exists_forall_sub_mem_ideal`,
+  the key to proving our theorem `exists_generator`.
+
+See [Karatarakis (2022)], for the definition of 'decomposition subgroup' in terms of
+  valuations.
+
+## Acknowledgements
+
+I would like to thank Prof. Buzzard's PhD students, Amelia and Jujian,
+for helping me to understand the math proof and to formalize it.
+
+-/
+
+open Classical
+
+section FiniteFrobeniusDef
+
+variable (A K L B : Type*) [CommRing A] [CommRing B]
+  [Algebra A B] [Field K] [Field L]
+  [IsDomain A] [IsDomain B]
+  [Algebra A K] [IsFractionRing A K]
+  [Algebra B L] [Algebra K L] [Algebra A L]
+  [IsScalarTower A B L] [IsScalarTower A K L]
+  [IsIntegralClosure B A L] [FiniteDimensional K L] [IsFractionRing B L]
+  [IsDedekindDomain A] [IsDedekindDomain B]
+
+/-- re-definition of `Ideal B` in terms of 'AKLB setup'. -/
+@[nolint unusedArguments] abbrev Ideal' (A K L B : Type*) [CommRing A]
```

Comment:
Naming things with `'` is rarely a good idea. Can you think up a name that gives the human a hint? Even `AKLB_Ideal`?

Reply:
Alternatively can you make it `private`??